### ISSUE TITLE
Clarification of training details

### ISSUE THREAD

**Issue by jwyang** [openai/CLIP#29]
Dear CLIP Authors,

Recently, I am reading the training details in the paper:

_To save additional memory, gradient checkpointing (Griewank & Walther, 2000; Chen et al., 2016), half-precision Adam statistics (Dhariwal et al., 2020), and half-precision stochastically rounded text encoder weights were used. The calculation of embedding
similarities was also sharded with individual GPUs computing only the subset of the pairwise similarities necessary for their local batch of embeddings._

I have two questions:

1. what does "half-precision stochastically rounded text encoder weights" mean? Is there any references?
2. I do not quite understanding "The calculation of embedding similarities was also sharded with individual GPUs". By saying "**sharded**" instead of "**shared**" Does this mean for each sample, the similarities to negative is only from the same gpu? Or you did a gather of features from other gpus and then compute the similarities between one sample to all.

thanks,
Jianwei

**Comment by jongwook** [comment]
1. When the fp32 encoder weights are cast to fp16, rather than deterministically rounding to the nearest fp16 point, stochastic rounding casts the value pseudorandomly to either ⌈x⌉ or ⌊x⌋ according to some probability. [This reference](https://hal.archives-ouvertes.fr/hal-02556997v2/document) mentions two modes for doing this, in (1.1) and (1.2).

2. Say there are N GPUs and each GPU gets M input pairs. When calculating similarities (and thus the logit scores across text inputs for each image) a GPU only needs to hold M image features, which needs to be matmul'ed with NM text features, resulting in an M×NM matrix. This computation is distributed (i.e. sharded) across N GPUs, and overall we have calculated NM×NM similarities across the GPUs. The loss we use is symmetric and the same happens w.r.t. text inputs.

**Comment by jwyang** [comment]
Hi, Jongwook,

thanks a lot for your quick reply! The explanations are very helpful to understand the training procedure!

**Comment by KeremTurgutlu** [comment]
> 2\. Say there are N GPUs and each GPU gets M input pairs. When calculating similarities (and thus the logit scores across text inputs for each image) a GPU only needs to hold M image features, which needs to be matmul'ed with NM text features, resulting in an M×NM matrix. This computation is distributed (i.e. sharded) across N GPUs, and overall we have calculated NM×NM similarities across the GPUs. The loss we use is symmetric and the same happens w.r.t. text inputs.

Thanks for the explanation, is there a sample code in pytorch on how to do that?
